File: backend/src/main/java/com/vvise/demo/security/OAuth2UserInfoFactory.java
package com.vvise.demo.security;

import com.vvise.demo.entity.User;

import java.util.Map;

public class OAuth2UserInfoFactory {

    public static OAuth2UserInfo getOAuth2UserInfo(String registrationId, Map<String, Object> attributes) {
        if (registrationId.equalsIgnoreCase(User.AuthProvider.GOOGLE.name())) {
            return new OAuth2UserInfo.GoogleOAuth2UserInfo(attributes);
        } else if (registrationId.equalsIgnoreCase(User.AuthProvider.GITHUB.name())) {
            return new OAuth2UserInfo.GithubOAuth2UserInfo(attributes);
        } else if (registrationId.equalsIgnoreCase(User.AuthProvider.MICROSOFT.name())) {
            return new OAuth2UserInfo.MicrosoftOAuth2UserInfo(attributes);
        } else {
            throw new RuntimeException("Login with " + registrationId + " is not supported.");
        }
    }

    public static User.AuthProvider getAuthProvider(String registrationId) {
        return User.AuthProvider.valueOf(registrationId.toUpperCase());
    }
}

File: backend/src/main/kotlin/com/vvise/auth/security/OAuth2UserInfoFactory.kt
package com.vvise.auth.security

import com.vvise.auth.entity.User

object OAuth2UserInfoFactory {

    fun getOAuth2UserInfo(registrationId: String, attributes: Map<String, Any>): OAuth2UserInfo {
        return when {
            registrationId.equals(User.AuthProvider.GOOGLE.name, ignoreCase = true) ->
                OAuth2UserInfo.GoogleOAuth2UserInfo(attributes)
            registrationId.equals(User.AuthProvider.GITHUB.name, ignoreCase = true) ->
                OAuth2UserInfo.GithubOAuth2UserInfo(attributes)
            registrationId.equals(User.AuthProvider.MICROSOFT.name, ignoreCase = true) ->
                OAuth2UserInfo.MicrosoftOAuth2UserInfo(attributes)
            else -> throw RuntimeException("Login with $registrationId is not supported.")
        }
    }

    fun getAuthProvider(registrationId: String): User.AuthProvider =
        User.AuthProvider.valueOf(registrationId.uppercase())
}
